### PR TITLE
Manual build mode support

### PR DESCRIPTION
**Pull request by grouma**
Closes https://github.com/dart-lang/build/issues/2314

- Create a public entrypoint for backend implementations of the daemon protocol.
  - Refer to `lib/daemon.dart`.
- Update client `connect` method to now take an optional `buildMode`.
  The default mode is auto in which builds will automatically occur on changes.
  The alternative mode is manual in which builds will only occur when triggered
  with the client `startBuild` method.
- Add enum of build modes to `constants.dart`.
- Update `build_runner` to support the latest `build_daemon`.
- Add tests for each of the modes

### CODE REVIEW

File: build_daemon/lib/change_provider.dart [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+abstract class ChangeProvider {}
+
+/// A provider of changes which are manually collected right before a build.
+abstract class ManualChangeProvider implements ChangeProvider {
+  Future<List<WatchEvent>> collectChanges();
+}
+
+/// A provider of changes which automatically cause builds to occur.
+abstract class AutoChangeProvider implements ChangeProvider {
+  /// If multiple files change together then they should be sent in the same
```
Comment (natebosch) [inline build_daemon/lib/change_provider.dart]: [nit] this isn't phrased as a doc comment.

https://dart.dev/guides/language/effective-dart/documentation#prefer-starting-function-or-method-comments-with-third-person-verbs
https://dart.dev/guides/language/effective-dart/documentation#do-start-doc-comments-with-a-single-sentence-summary

File: build_daemon/lib/change_provider.dart [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:watcher/src/watch_event.dart';
+
+abstract class ChangeProvider {}
```
Comment (natebosch) [inline build_daemon/lib/change_provider.dart]: I'd be more comfortable if both methods were defined here and we have a doc comment that says it's fine for one method to always return "empty" if the other method encompasses all changes.

In the long term we can also consider going through the `collectChanges` path even if the build was triggered by something on the `changes` stream with the intent of hypothetical use cases like "we auto rebuild if you edit in `lib/`, but if you edit somewhere else you need to manually trigger a build"

File: build_daemon/lib/change_provider.dart (hunk 1 above)
Comment (grouma) [reply build_daemon/lib/change_provider.dart]: Done.

File: build_daemon/lib/change_provider.dart (hunk 2 above)
Comment (grouma) [reply build_daemon/lib/change_provider.dart]: Done.

**Pull request merged**